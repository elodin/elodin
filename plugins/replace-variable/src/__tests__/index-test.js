import { traverse, parse } from '@elodin/core'

import replaceVariable from '../index'

describe('Replacing variables', () => {
  it('should correctly replace variables in declarations', () => {
    const file = `style Button { paddingLeft: 10 paddingRight: $spacingXL }`

    const ast = parse(file).ast

    expect(
      traverse(ast, [
        replaceVariable({
          variables: {
            spacingL: 10,
            spacingXL: 20,
          },
        }),
      ])
    ).toMatchSnapshot()
  })
  it('should correctly replace variables in functions', () => {
    const file = `style Button { paddingLeft: 10 paddingRight: add($spacingXL 1) }`

    const ast = parse(file).ast

    expect(
      traverse(ast, [
        replaceVariable({
          variables: {
            spacingL: 10,
            spacingXL: 20,
          },
        }),
      ])
    ).toMatchSnapshot()
  })

  it('should correctly replace variables in conditions', () => {
    const file = `style Button { paddingLeft: 10 [@viewportWidth>=$fromM] { paddingLeft: 20 } }`

    const ast = parse(file).ast

    expect(
      traverse(ast, [
        replaceVariable({
          variables: {
            fromM: 480,
          },
        }),
      ])
    ).toMatchSnapshot()
  })

  it('should work with custom selectors', () => {
    const file = `style Button { paddingLeft: 10 backgroundColor: $colors_primary }`

    const ast = parse(file).ast

    expect(
      traverse(ast, [
        replaceVariable({
          variables: {
            colors: {
              primary: 'red',
              secondary: 'blue',
            },
          },

          selector: (variables, property) =>
            property
              .split('_')
              .reduce((out, prop) => (out ? out[prop] : undefined), variables),
        }),
      ])
    ).toMatchSnapshot()
  })
})
