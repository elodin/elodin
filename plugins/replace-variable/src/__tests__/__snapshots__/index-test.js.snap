// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Replacing variables should correctly replace variables in conditions 1`] = `
Object {
  "body": Array [
    Object {
      "body": Array [
        Object {
          "comments": Array [],
          "dynamic": false,
          "property": "paddingLeft",
          "raw": false,
          "type": "Declaration",
          "value": Object {
            "negative": false,
            "type": "Integer",
            "value": 10,
          },
        },
        Object {
          "body": Array [
            Object {
              "comments": Array [],
              "dynamic": false,
              "property": "paddingLeft",
              "raw": false,
              "type": "Declaration",
              "value": Object {
                "negative": false,
                "type": "Integer",
                "value": 20,
              },
            },
          ],
          "comments": Array [],
          "dynamic": false,
          "operator": ">=",
          "property": Object {
            "environment": true,
            "type": "Variable",
            "value": "viewportWidth",
          },
          "type": "Conditional",
          "value": Object {
            "type": "Integer",
            "value": 480,
          },
        },
      ],
      "comments": Array [],
      "name": "Button",
      "type": "Style",
    },
  ],
  "type": "File",
}
`;

exports[`Replacing variables should correctly replace variables in declarations 1`] = `
Object {
  "body": Array [
    Object {
      "body": Array [
        Object {
          "comments": Array [],
          "dynamic": false,
          "property": "paddingLeft",
          "raw": false,
          "type": "Declaration",
          "value": Object {
            "negative": false,
            "type": "Integer",
            "value": 10,
          },
        },
        Object {
          "comments": Array [],
          "dynamic": false,
          "property": "paddingRight",
          "raw": false,
          "type": "Declaration",
          "value": Object {
            "type": "Integer",
            "value": 20,
          },
        },
      ],
      "comments": Array [],
      "name": "Button",
      "type": "Style",
    },
  ],
  "type": "File",
}
`;

exports[`Replacing variables should correctly replace variables in functions 1`] = `
Object {
  "body": Array [
    Object {
      "body": Array [
        Object {
          "comments": Array [],
          "dynamic": false,
          "property": "paddingLeft",
          "raw": false,
          "type": "Declaration",
          "value": Object {
            "negative": false,
            "type": "Integer",
            "value": 10,
          },
        },
        Object {
          "comments": Array [],
          "dynamic": true,
          "property": "paddingRight",
          "raw": false,
          "type": "Declaration",
          "value": Object {
            "callee": "add",
            "params": Array [
              Object {
                "type": "Integer",
                "value": 20,
              },
              Object {
                "negative": false,
                "type": "Integer",
                "value": 1,
              },
            ],
            "type": "FunctionExpression",
          },
        },
      ],
      "comments": Array [],
      "name": "Button",
      "type": "Style",
    },
  ],
  "type": "File",
}
`;

exports[`Replacing variables should work with custom selectors 1`] = `
Object {
  "body": Array [
    Object {
      "body": Array [
        Object {
          "comments": Array [],
          "dynamic": false,
          "property": "paddingLeft",
          "raw": false,
          "type": "Declaration",
          "value": Object {
            "negative": false,
            "type": "Integer",
            "value": 10,
          },
        },
        Object {
          "comments": Array [],
          "dynamic": false,
          "property": "backgroundColor",
          "raw": false,
          "type": "Declaration",
          "value": Object {
            "type": "Identifier",
            "value": "red",
          },
        },
      ],
      "comments": Array [],
      "name": "Button",
      "type": "Style",
    },
  ],
  "type": "File",
}
`;
